src/stdlib_linalg_least_squares.fypp

```diff
@@ -0,0 +1,215 @@
+#:include "common.fypp"
+#:set RC_KINDS_TYPES = REAL_KINDS_TYPES + CMPLX_KINDS_TYPES
+#:set RHS_SUFFIX = ["one","many"]
+#:set RHS_SYMBOL = [ranksuffix(r) for r in [1,2]]
+#:set RHS_EMPTY  = [emptyranksuffix(r) for r in [1,2]]
+#:set ALL_RHS    = list(zip(RHS_SYMBOL,RHS_SUFFIX,RHS_EMPTY))
+submodule (stdlib_linalg) stdlib_linalg_least_squares
+!! Least-squares solution to Ax=b
+     use stdlib_linalg_constants
+     use stdlib_linalg_lapack, only: gelsd, stdlib_ilaenv
+     use stdlib_linalg_state, only: linalg_state_type, linalg_error_handling, LINALG_ERROR, &
+         LINALG_INTERNAL_ERROR, LINALG_VALUE_ERROR
+     implicit none(type,external)
+     
+     character(*), parameter :: this = 'lstsq'
+
+     contains
+
+     #:for rk,rt,ri in RC_KINDS_TYPES
+     #:if rk!="xdp"
+     ! Workspace needed by gesv
+     elemental subroutine ${ri}$gesv_space(m,n,nrhs,lrwork,liwork,lcwork)
+         integer(ilp), intent(in) :: m,n,nrhs
+         integer(ilp), intent(out) :: lrwork,liwork,lcwork
+
+         integer(ilp) :: smlsiz,mnmin,nlvl
+
+         mnmin = min(m,n)
+
+         ! Maximum size of the subproblems at the bottom of the computation (~25)
+         smlsiz = stdlib_ilaenv(9,'${ri}$gelsd',' ',0,0,0,0)
+
+         ! The exact minimum amount of workspace needed depends on M, N and NRHS. As long as LWORK is at least
+         nlvl = max(0, ilog2(mnmin/(smlsiz+1))+1)
+
+         ! Real space
+         #:if rt.startswith('complex')
+         lrwork = 10*mnmin+2*mnmin*smlsiz+8*mnmin*nlvl+3*smlsiz*nrhs+max((smlsiz+1)**2,n*(1+nrhs)+2*nrhs)
+         #:else
+         lrwork = 12*mnmin+2*mnmin*smlsiz+8*mnmin*nlvl+mnmin*nrhs+(smlsiz+1)**2
+         #:endif
+         lrwork = max(1,lrwork)
+
+         ! Complex space
+         lcwork = 2*mnmin + nrhs*mnmin
+
+         ! Integer space
+         liwork = max(1, 3*mnmin*nlvl+11*mnmin)
+
+         ! For good performance, the workspace should generally be larger.
+         lrwork = ceiling(1.25*lrwork,kind=ilp)
+         lcwork = ceiling(1.25*lcwork,kind=ilp)
+         liwork = ceiling(1.25*liwork,kind=ilp)
+
+     end subroutine ${ri}$gesv_space
+
+     #:endif
+     #:endfor
+
+     #:for nd,ndsuf,nde in ALL_RHS
+     #:for rk,rt,ri in RC_KINDS_TYPES
+     #:if rk!="xdp"
+
+     ! Compute the least-squares solution to a real system of linear equations Ax = b
+     module function stdlib_linalg_${ri}$_lstsq_${ndsuf}$(a,b,cond,overwrite_a,rank,err) result(x)
```

Comment:
The default functional style here is chosen as it is in line with the approach taken by the most used libraries. I definitely support having one more interface using a `subroutine`. In this case, what should be called like? Do you think it could be a good idea to open a separate PR with related discussion on it?

Reply:
>In this case, what should be called like?

I'm not sure if there is already a convention within stdlib regarding an implementation having both a function and a subroutine interfaces simultaneously ... could be something like `slstsq` ? too many letters I feel... when working on `to_num` I used `*_base` where base where all subroutines, which are called by the function interface, so another idea:  `lstsq_base` ?

>Do you think it could be a good idea to open a separate PR with related discussion on it?
That can be a good solution to avoid holding this PR ;)

Reply:
I was trying to see if I could cook a quick subroutine interface. However, we also need to decide what to do with internal allocation (can be significant): 

https://github.com/fortran-lang/stdlib/blob/90f553f0b28c0139ca069524be7cd8e37197ab6d/src/stdlib_linalg_least_squares.fypp#L156-L162

As long as this is always allocated in place, re-using `x` will not lead to significant performance benefits. So yes, I would defer this to a separate PR!

Reply:
>I'm not sure if there is already a convention within stdlib regarding an implementation having both a function and a subroutine interfaces simultaneously

I am not aware of such conventions. However, I am in favour of providing both functions and subroutines when it is possible, and especially when large arrays might be involved..

Reply:
I like your idea @jalvesz. To save us from the hassle of doing this for every single function, should we envision a storage space DT meant to work with *all* linear algebra functions? Something like: 

```
type :: linalg_storage
   integer(ilp), allocatable :: iwork(:)
   real(sp), allocatable :: rwork_sp(:) 
   real(dp), allocatable :: rwork_dp(:)
   real(qp), allocatable :: rwork_qp(:)
   complex(sp), allocatable :: cwork_sp(:) 
   complex(dp), allocatable :: cwork_dp(:)
   complex(qp), allocatable :: cwork_qp(:)
   ...
end type
```

There is no way to avoid embedding all possible datatypes, but we could at least simplify the initialization: 

```
type(linalg_storage) :: work
work = linalg_storage('solve',A,b)
work = linalg_storage('lstsq',A,b)
...
```

Reply:
You could split it by kind:
```Fortran
type, abstract :: linalg_storage
   integer(ilp), allocatable :: iwork(:)
   integer(ilp), allocatable :: pivot(:)
end type

type, extends(linalg_storage) :: linalg_storage_sp
   real(sp), allocatable :: diagonals(:) 
   real(sp), allocatable :: work(:) 
end type

type, extends(linalg_storage) :: linalg_storage_dp
   real(dp), allocatable :: diagonals(:) 
   real(dp), allocatable :: work(:) 
end type

type, extends(linalg_storage) :: linalg_storage_csp
   complex(sp), allocatable :: diagonals(:) 
   complex(sp), allocatable :: work(:) 
end type

type, extends(linalg_storage) :: linalg_storage_cdp
   complex(sp), allocatable :: diagonals(:) 
   complex(sp), allocatable :: work(:) 
end type
```

And for the user he would need to select according to the working precession:
```
type(linalg_storage_sp) :: work
work = linalg_storage('solve',A,b) !> or  linalg_storage('solve',num_rows,num_cols,num_rhs) ?
work = linalg_storage('lstsq',A,b) !> or  linalg_storage('solve',num_rowsnum_cols,num_rhs) ?
```

Reply:
Respectfully, I believe this looks a bit overengineered. I'm not sure if the user would be happy to check what version of the derived type they're using every time. Keep in mind that internally, the sizes of all necessary arrays will always be checked. So if the user tries to submit an invalid storage variable to a problem, an error will be returned. 

However, I also agree with you that having so many `allocatable` arrays inside it is also pretty dumb/non-optimal (PS: `complex` cases need 3: real, integer, complex storage). The best way to me would be to have a parameterized derived type. We may try it as we don't need type-bound procedures, but I don't know if compiler support is wide enough: 

```fortran
type :: linalg_storage(wp)
   integer, kind :: wp
   integer(ilp), allocatable :: pivot(:)
   real(wp), allocatable :: diagonals(:) 
   real(wp), allocatable :: rwork(:) 
   complex(wp), allocatable :: cwork(:) 
end type
```

So at the end of the day, maybe the big derived type is the best tradeoff?

Reply:
Right, the PDT seems like the ideal solution, I'm also worried about the support in compilers, but keepings its use plain should allow for a broad use. 
>So at the end of the day, maybe the big derived type is the best tradeoff?

Probably.